drv/stm32h7-spi-server/src/main.rs

```diff
@@ -360,79 +364,85 @@ impl ServerImpl {
+        let mut tx_permits = 16;
+
+        // We monitor our overall progress based on bytes _received,_ since
+        // every TX has a corresponding RX.
+        let mut rx_count = 0;
+        // We also keep track of bytes TX'd, though, to make sure we let the TX
+        // FIFO empty at the end of transmission.
+        let mut tx_count = 0;
+
         // While work remains, we'll attempt to move up to one byte
         // in each direction, sleeping if we can do neither.
-        while tx.is_some() || rx.is_some() {
+        while rx_count != overall_len {
             // Entering RECV to check for interrupts is not free, so
             // we only do it if we've filled the TX FIFO and emptied
             // the RX and repeating this loop would just burn power
             // and CPU. If there is any potential value to repeating
             // the loop immediately, we'll set this flag.
             let mut made_progress = false;
 
-            if let Some((tx_data, tx_pos)) = &mut tx {
-                while self.spi.can_tx_frame() {
-                    // If our position is less than our tx len,
-                    // transfer a byte from caller to TX FIFO --
-                    // otherwise put a dummy byte on the wire
-                    let byte: u8 = if *tx_pos < src_len {
-                        tx_data
-                            .read_at(usize::from(*tx_pos))
-                            .ok_or(RequestError::went_away())?
-                    } else {
-                        0u8
-                    };
-
-                    ringbuf_entry!(Trace::Tx(*tx_pos, byte));
+            // If there are things to transmit in the first place...
+            if let Some(tx_reader) = &mut tx {
+                // ...and if we're not going to blow either FIFO...
+                while tx_count < overall_len
+                    && tx_permits > 0
+                    && self.spi.can_tx_frame()
+                {
+                    // If we read off the end, or if the client goes away, we'll
+                    // substitute zero. This allows the TX to be shorter than RX
+                    // and get padded.
+                    let byte = tx_reader.read().unwrap_or(0);
+
+                    ringbuf_entry!(Trace::Tx(byte));
                     self.spi.send8(byte);
-                    *tx_pos += 1;
+                    tx_permits -= 1;
+                    tx_count += 1;
                     made_progress = true;
-
-                    if *tx_pos == overall_len {
-                        // We continue to transmit beyond our src_len
-                        // (transmitting dummy bytes, above), but once we
-                        // hit overall_len we are done transmitting.
-                        tx = None;
-                        break;
-                    }
+                }
+                if tx_count == overall_len {
+                    // Optimization: stop feeding the FIFO and don't repeat
```

Comment:
It would be helpful to add a comment that disabling CAN interrupts _also_ protects us from an IRQ storm in the weird case where the hardware gets stuck (until we find the actual root cause there).

Reply:
Done!